Spigot-Server-Patches/0418-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,251 @@
++                this.a(j, b0, this.c(b0), b0 <= i - 2, i); //Paper ticket level pp-view-dist
+             }
+ 
+             this.e = i;
+         }
+ 
+-        private void a(long i, int j, boolean flag, boolean flag1) {
++        private void a(long i, int j, boolean flag, boolean flag1, int vd) { //Paper ticket level pp-view-dist
+             if (flag != flag1) {
+                 Ticket<?> ticket = new Ticket<>(TicketType.PLAYER, ChunkMapDistance.b, new ChunkCoordIntPair(i), ChunkMapDistance.this.currentTick);
+ 
+                 if (flag1) {
+                     ChunkMapDistance.this.j.a(ChunkTaskQueueSorter.a(() -> { // Craftbukkit - decompile error
+                         ChunkMapDistance.this.m.execute(() -> {
+-                            if (this.c(this.c(i))) {
++                            if (this.c(i) <= vd) { //Paper ticket level pp-view-dist
+                                 ChunkMapDistance.this.addTicket(i, ticket);
+                                 ChunkMapDistance.this.l.add(i);
+                             } else {
+@@ -430,10 +524,33 @@ public abstract class ChunkMapDistance {
+                             }
+ 
+                         });
+-                        this.a(i, k, this.c(j), this.c(k));
++                    }
++
++                    //Paper start ticket level pp-view-dist
++                    byte f = ChunkMapDistance.this.update.remove(i);
++                    if(f == 1) {
++                        this.a(i, k, false, true, Integer.MAX_VALUE);
++                    } else if (f == -1) {
++                        this.a(i, k, true, false, Integer.MAX_VALUE);
+                     }
+                 }
+ 
++                ObjectIterator<it.unimi.dsi.fastutil.longs.Long2ByteMap.Entry> it = ChunkMapDistance.this.update.long2ByteEntrySet().iterator();
++
++                while(it.hasNext()) {
++                    it.unimi.dsi.fastutil.longs.Long2ByteMap.Entry e = it.next();
++                    long i = e.getLongKey();
++                    byte f = e.getByteValue();
++                    if(f == 1) {
++                        this.a(i, this.c(i), false, true, Integer.MAX_VALUE);
++                    } else if (f == -1) {
++                        this.a(i, this.c(i), true, false, Integer.MAX_VALUE);
++                    }
++                }
++
++                ChunkMapDistance.this.update.clear();
++                //Paper end
++
+                 this.g.clear();
+             }
+ 
+diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
+index cd73cde8..2186a81e 100644
+--- a/src/main/java/net/minecraft/server/EntityHuman.java
++++ b/src/main/java/net/minecraft/server/EntityHuman.java
+@@ -80,6 +80,16 @@ public abstract class EntityHuman extends EntityLiving {
+     public String spawnWorld = "";
+     public int oldLevel = -1;
+ 
++    //Paper start ticket level pp-view-dist
++    private int viewDistance;
```

Comment:
we're still lacking correct control over sending chunks to players

Reply:
yes, this is base level logic as stated. that is in progress.

Reply:
do not mark this is as resolved if it is not resolved

Reply:
it has been stated very clearly that this is base level logic, and that higher level etc. is in progress. this is listed on the todo list and you have mentioned this above.